pypose/module/ukf.py

```diff
@@ -0,0 +1,319 @@
+        4. Perform UKF prediction. Note that estimation error becomes smaller with more steps.
+
+        >>> for i in range(T - 1):
+        ...     w = q * torch.randn(N) # transition noise
+        ...     v = r * torch.randn(N) # observation noise
+        ...     states[i+1], observ[i] = model(states[i] + w, inputs[i])
+        ...     estim[i+1], P[i+1] = ukf(estim[i], observ[i] + v, inputs[i], P[i], Q, R)
+        ... print('Est error:', (states - estim).norm(dim=-1))
+        Est error: tensor([8.8161, 9.0322, 5.4756, 2.2453, 0.9141])
+
+    Warning:
+        Don't introduce noise in ``System`` methods ``state_transition`` and ``observation``
+        for filter testing, as those methods are used for automatically linearizing the system
+        by the parent class ``pypose.module.System``, unless your system model explicitly
+        introduces noise.
+
+    Note:
+        Implementation is based on Section 14.3 of this book
+
+        * Dan Simon, `Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches
+          <https://onlinelibrary.wiley.com/doi/epdf/10.1002/0470045345.fmatter>`_,
+          Cleveland State University, 2006
+    '''
+
+    def __init__(self, model, Q=None, R=None):
+        super().__init__()
+        self.set_uncertainty(Q=Q, R=R)
+        self.model = model
+
+    def compute_weight(self):
+        r'''
+        compute ukf weight
+
+        Return:
+        out (:obj:`int`): the ukf weight.
+        '''
+        return 1 / (2 * self.dim)
+
+    def compute_sigma(self, x, u, P, C, D, c2):  # compute sigma point
+        r'''
+        compute sigma point and Observation
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            u (:obj:`Tensor`): system input at current step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+            C (:obj:`Tensor`): The output matrix of LTI system.
+            D (:obj:`Tensor`): The observation matrix of LTI system,
+            c2 (:obj:`Tensor`): The constant output of LTI system.
+
+        Return:
+            list of :obj:`Tensor`: sigma point and Oerservation
+        '''
+        # gather index
+        index_repeat = tuple(
+            torch.cat([torch.tensor([2], dtype=torch.int64), torch.ones(self.dim - 1, dtype=torch.int64)]).numpy())
+        index_weight = torch.arange(self.dim, device=P.device, dtype=P.dtype).unsqueeze(1)
+        index = torch.ones(self.dim, self.dim, device=P.device, dtype=P.dtype)
+
+        index_gather = torch.as_tensor(index * index_weight, dtype=torch.int64, device=P.device).repeat(
+            index_repeat).unsqueeze(1).reshape(2 * self.dim, 1, self.dim)
+
+        # compute root of P
+        repeat_dim = tuple(torch.cat([torch.tensor([self.dim * 2]), torch.tensor(P.shape)]).numpy())  # repeat dim
+        p_expand = P.expand(repeat_dim)
+        np_expand = self.dim * p_expand  # calculate np
+        np_expand_sqrt = msqrt(np_expand)  # square root of np
```

Comment:
`P` is always positive definite, thus we may be able to use [Cholesky decomposition](https://pytorch.org/docs/stable/generated/torch.linalg.cholesky.html) for matrix square root, which could be faster. You may want to test for speed comparison.

Reply:
https://github.com/HongLouyemeng/pypose/blob/e28bb3a81378046c6cd46e642a41f8daab4b15ed/pypose/basics/ops.py#L214-L273, Cholesky in pytorch will make the ukf unstable, I have found a method, but its covariance will make the eigenvalue negative in **Bicycle Model**,but it will not occur based on **contour_integral** or **pade**.

Reply:
Reference:
  [1][https://iopscience.iop.org/article/10.1209/0295-5075/78/12001](url)
  [2][http://eprints.maths.manchester.ac.uk/1136/1/hht08.pdf](url)

Reply:
Cholesky decomposition  is faster than Padé approximant.